Ff153 Add IDBRecord (#44540)
mdn/content: files/en-us/web/api/idbrecord/index.md

---
title: IDBRecord
slug: Web/API/IDBRecord
page-type: web-api-interface
browser-compat: api.IDBRecord
---

{{APIRef("IndexedDB")}} {{AvailableInWorkers}}

The **`IDBRecord`** interface of the [IndexedDB API](/en-US/docs/Web/API/IndexedDB_API) represents a snapshot of a single record in an {{domxref("IDBObjectStore")}} or {{domxref("IDBIndex")}}.

A request for records using {{domxref("IDBObjectStore.getAllRecords()")}} or {{domxref("IDBIndex.getAllRecords()")}} returns an {{domxref("IDBRequest")}} instance.

> On success the returned object's {{domxref("IDBRequest.result", "result")}} property is populated with an array of `IDBRecord` instances.

Reviewer: ```suggestion
On success, the returned object's {{domxref("IDBRequest.result", "result")}} property is populated with an array of `IDBRecord` instances.
```

## Instance properties

- `key` {{ReadOnlyInline}}

>   - : A value representing the record's key.

Author: ```suggestion
  - : A value representing the record's secondary key.
```

    For an object store record this will be the same as `primaryKey`.

>     For an index record it will be the record's key within the index.

Reviewer: ```suggestion
    For an object store record, this will be the same as `primaryKey`.
    For an index record, it will be the record's key within the index.
```

- `primaryKey` {{ReadOnlyInline}}

>   - : A value representing the key of the record in the index's associated {{domxref("IDBObjectStore")}}.

Reviewer: This is index-specific description, whereas `key` distinguishes between object store record and index record.

Should `primaryKey` do the same?
or can it be simplified to: "A value representing the record's primary key."

Author: Yes, it can be simplified. Though it is worth noting that it is the key used in the store.

```suggestion
  - : A value representing the record's primary key.
    This key is used to represent the record in the {{domxref("IDBObjectStore")}}.
```

- `value` {{ReadOnlyInline}}
  - : A value representing the record's value.

## Instance methods

_None._

## Example

In the following code snippet, we open a database asynchronously and make a request; `onerror` and `onsuccess` functions are included to handle the success and error cases.

> For a full working example, see our [To-do Notifications](https://github.com/mdn/dom-examples/tree/main/to-do-notifications) app ([view example live](https://mdn.github.io/dom-examples/to-do-notifications/).)

Reviewer: This feels like a mismatch for the code snippet below. The code does not have the functions mentioned.

The "To-do Notifications" app does not have occurrences of `getAllRecords`.

Author: Absolutely right. I'll fix.

Reviewer: Thanks! Much better now

```js
const query = IDBKeyRange.lowerBound("myKey", true);
const objectStore = transaction.objectStore("contactsList");

const myRecords = (objectStore.getAllRecords({
  query,

>   count: "100",

Reviewer: `count` is defined as a [number](https://developer.mozilla.org/en-US/docs/Web/API/IDBIndex/getAllRecords#count), yet our code examples for both `getAllRecords()` and now here as well show it as a string. Is that correct?
For comparison, the linked explainer at https://microsoftedge.github.io/Demos/idb-getallrecords/ has `count` value as a number data type.

Author: Yes, copied from getAllRecords in the IDBIndex, which is also wrong.

```suggestion
  count: 100,
```

A request for records using {{domxref("IDBObjectStore.getAllRecords()")}} or {{domxref("IDBIndex.getAllRecords()")}} returns an {{domxref("IDBRequest")}} instance.
On success, the returned object's {{domxref("IDBRequest.result", "result")}} property is populated with an array of `IDBRecord` instances.

## Instance properties

- `key` {{ReadOnlyInline}}
  - : A value representing the record's secondary key.
    For an object store record, this will be the same as `primaryKey`.
    For an index record, it will be the record's key within the index.
- `primaryKey` {{ReadOnlyInline}}
  - : A value representing the record's primary key.
    This key is used to represent the record in the {{domxref("IDBObjectStore")}}.
- `value` {{ReadOnlyInline}}
  - : A value representing the record's value.

## Instance methods

_None._

## Example

In the following code snippet, we open a database asynchronously and make a request; `onerror` and `onsuccess` functions are included to handle the success and error cases.

> For a full working example, see our [To-do Notifications](https://github.com/mdn/dom-examples/tree/main/to-do-notifications) app ([view example live](https://mdn.github.io/dom-examples/to-do-notifications/).)

Author: ```suggestion
## Example

### Basic usage

This code snippet queries an `IDBObjectStore` for up to 100 records whose keys come after `"myKey"`, with results to be sorted in reverse order.
The results are returned via the `onsuccess` callback in `myRecords.result`, which contains an array of `IDBRecord` instances.
Note that because this is a query on `IDBObjectStore`, the `key` and `primaryKey` in each record have the same value.
```

---
title: IDBRecord
slug: Web/API/IDBRecord
page-type: web-api-interface
browser-compat: api.IDBRecord
---

{{APIRef("IndexedDB")}} {{AvailableInWorkers}}

The **`IDBRecord`** interface of the [IndexedDB API](/en-US/docs/Web/API/IndexedDB_API) represents a snapshot of a single record in an {{domxref("IDBObjectStore")}} or {{domxref("IDBIndex")}}.

A request for records using {{domxref("IDBObjectStore.getAllRecords()")}} or {{domxref("IDBIndex.getAllRecords()")}} returns an {{domxref("IDBRequest")}} instance.
On success, the returned object's {{domxref("IDBRequest.result", "result")}} property is populated with an array of `IDBRecord` instances.

## Instance properties

- `key` {{ReadOnlyInline}}
  - : A value representing the record's secondary key.
    For an object store record, this will be the same as `primaryKey`.
    For an index record, it will be the record's key within the index.
- `primaryKey` {{ReadOnlyInline}}
  - : A value representing the record's primary key.
    This key is used to represent the record in the {{domxref("IDBObjectStore")}}.
- `value` {{ReadOnlyInline}}
  - : A value representing the record's value.

## Instance methods

_None._

> ## Example

Reviewer: ```suggestion
## Examples
```

### Basic usage

This code snippet queries an `IDBObjectStore` for up to 100 records whose keys come after `"myKey"`, with results to be sorted in reverse order.
The results are returned via the `onsuccess` callback in `myRecords.result`, as an array of `IDBRecord` instances.
Note that because this is a query on an `IDBObjectStore`, the `key` and `primaryKey` in each record have the same value.

```js
const query = IDBKeyRange.lowerBound("myKey", true);
const objectStore = transaction.objectStore("contactsList");

const myRecords = (objectStore.getAllRecords({
  query,
  count: 100,
  direction: "prev",

> }).onsuccess = (event) => {

Reviewer: - you may want to consider splitting this code into two statements, that is, one for assigning to `myRecords` and a separate one for `myRecords.onsuccess`

- `transaction` in l41 is not defined here. Maybe we can borrow the transaction setup lines of code from https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/get#examples?

I realize both these issues exist in the examples we have on the two `getAllRecords()` pages (Looking at the history, I had done the editorial review of those methods when they were first added in https://github.com/mdn/content/pull/41398)

Author: Thanks @dipikabh 

- I have split all the exmaples as you suggested - its a much cleaner pattern. 
- Also added as event listeners which is how you are supposed to do these things. Modified in all three examples in same way (modulo the difference in index vs object store).
- I defined `transaction` but NOT the database that is needed for it (at some point you get to where all the infrastructure means you no longer have an example that is "basic usage). In any case, I fixed the issue that this was not defined by noting explicitly that this was not done, rather than just not mentioning it. 
- I have similarly updated the examples in the IDBRequest and record to use the addEventListener. Didn't take it further than that for those. 

Updates all in https://github.com/mdn/content/pull/44540/changes/ddc1b173ec572bffc5c4716f195c269a7aeba58d

I think it is good. Leaving in case you want to scan it again.

Reviewer: the updated code and prose are looking really good, great improvement. thanks!
